oauth2client/flask_util.py

```diff
@@ -0,0 +1,402 @@
+            server_csrf = session['google_oauth2_state']
+            code = request.args['code']
+        except KeyError:
+            return 'Invalid request', 400
+
+        if client_csrf != server_csrf:
+            return 'Invalid request state', 400
+
+        flow = self._make_flow()
+
+        # Exchange the auth code for credentials.
+        try:
+            credentials = flow.step2_exchange(code)
+        except FlowExchangeError as e:
+            current_app.logger.exception(e)
+            return 'An error occurred: %s' % e, 400
+
+        # Save the credentials to the storage.
+        self.storage.put(credentials)
+
+        if self.authorize_callback:
+            self.authorize_callback(credentials)
+
+        return redirect(session.pop('google_oauth2_return', '/'))
+
+    @property
+    def credentials(self):
+        ctx = _app_ctx_stack.top
+
+        if not hasattr(ctx, 'google_oauth2_credentials'):
+            ctx.google_oauth2_credentials = self.storage.get()
+
+        return ctx.google_oauth2_credentials
+
+    def has_credentials(self):
+        return self.credentials and not self.credentials.invalid
+
+    @property
+    def email(self):
+        """Returns the user's email address.
+
+        The email address is provided by the current credentials' id_token. The
+        use can change their email, it's more appropriate to use user_id for
+        a unique, user-specific identifier.
+        """
+        if not self.credentials:
+            return None
+        try:
+            return self.credentials.id_token['email']
+        except KeyError:
+            current_app.logger.error(
+                'Invalid id_token %s', self.credentials.id_token)
+
+    @property
+    def user_id(self):
+        """Returns the user's email address.
+
+        The id is provided by the current credentials' id_token.
+        """
+        if not self.credentials:
+            return None
+        try:
+            return self.credentials.id_token['sub']
+        except KeyError:
+            current_app.logger.error(
+                'Invalid id_token %s', self.credentials.id_token)
+
+    def authorize_url(self, return_url, **kwargs):
+        """Creates a URL that will walk the user through the authorization flow.
+
+        Returns the user to the specified return_url once authorization is
+        complete. Any kwargs are passed into the flow constructor.
+        """
+        session['google_oauth2_flow_kwargs'] = kwargs
+        return url_for('oauth2.authorize', return_url=return_url)
+
+    def required(self, decorated_function=None, **decorator_kwargs):
+        """Decorator to require OAuth2 credentials for a view.
+
+        If credentials are not available for the current user, then they will
+        be redirected to the authorization flow. Once complete, the user will
+        be redirected back to the original page.
+        """
+        def curry_wrapper(decorated_function):
```

Comment:
Suggest using a different name the the one already in the scope (`decorated_function` is in the `required` function scope since in the args)